task/net/src/server.rs

```diff
@@ -240,3 +233,343 @@ impl<T: NetServer> idl::InOrderNetImpl for T {
+            // Bind sockets to their ports.
+            for (&h, port) in zip(&socket_handles, generated::SOCKET_PORTS) {
+                iface
+                    .get_socket::<UdpSocket<'_>>(h)
+                    .bind((ipv6_addr, port))
+                    .unwrap_lite();
+            }
+
+            vlan_state
+                .push(VLanState {
+                    socket_handles,
+                    iface,
+                })
+                .unwrap_lite();
+
+            // Increment the MAC and IP addresses so that each VLAN has
+            // a unique address.
+            ipv6_addr.0[15] += 1;
+            mac.0[5] += 1;
+        }
+
+        Self {
+            eth,
+            client_waiting_to_send: [false; SOCKET_COUNT],
+            vlan_state: vlan_state.into_array().unwrap_lite(),
+            bsp,
+            mac: start_mac,
+        }
+    }
+
+    pub(crate) fn poll(&mut self, t: u64) -> smoltcp::Result<crate::Activity> {
+        let t = smoltcp::time::Instant::from_millis(t as i64);
+        // Do not be tempted to use `Iterator::any` here, it short circuits and
+        // we really do want to poll all of them.
+        let mut ip = false;
+        let mut mac_rx = false;
+        for vlan in &mut self.vlan_state {
+            ip |= vlan.iface.poll(t)?;
+            // Test and clear our receive activity flag.
+            mac_rx |= vlan.iface.device().read_and_clear_activity_flag();
+        }
+
+        Ok(crate::Activity { ip, mac_rx })
+    }
+
+    /// Iterate over sockets, waking any that can do work.  A task can do work
+    /// if all of the (internal) VLAN sockets can receive a packet, since
+    /// we don't know which VLAN it will write to.
+    pub fn wake_sockets(&mut self) {
+        for i in 0..SOCKET_COUNT {
+            if (0..N).any(|v| {
+                let want_to_send = self.client_waiting_to_send[i];
+                let socket = self.get_socket_mut(i, v).unwrap();
+                socket.can_recv() || (want_to_send && socket.can_send())
+            }) {
+                let (task_id, notification) = generated::SOCKET_OWNERS[i];
+                let task_id = sys_refresh_task_id(task_id);
+                sys_post(task_id, notification);
+            }
+        }
+    }
+
+    pub fn wake(&self) {
+        self.bsp.wake(self.eth)
+    }
+
+    fn get_handle(
+        &self,
+        index: usize,
+        vlan_index: usize,
+    ) -> Option<SocketHandle> {
+        self.vlan_state
+            .get(vlan_index)?
+            .socket_handles
+            .get(index)
+            .cloned()
+    }
+
+    /// Gets the socket `index`. If `index` is out of range, returns
+    /// `BadMessage`. Panics if `vlan_index` is out of range, which should
+    /// never happen (because messages with invalid VIDs are dropped in
+    /// RxRing).
+    ///
+    /// Sockets are currently assumed to be UDP.
+    pub(crate) fn get_socket_mut(
```

Comment:
Now that we've consolidated `iface` and `socket_handles` into the `VLanState`, it seems like `get_handle` and `get_socket_mut` could be functions on that struct, e.g.
```rust
impl VLanState
    fn get_handle(&self, index: usize) -> Option<SocketHandle> {
        self.socket_handles.get(index).cloned()
    }
    fn get_socket_mut(
        &mut self,
        index: usize,
    ) -> Option<&mut UdpSocket<'static>> {
        self.iface.get_socket::<UdpSocket<'_>>(self.get_handle(index)?)
    }
}
```

This removes repeated indexing of `self.vlan_state`

Reply:
If only rustfmt would let those routines be so vertically compact. :-)

The result of this is better than you implied -- not only does it remove repeated indexing, it actually removes _the indexing_ in most cases. e.g.

```diff
         // Iterate over all of the per-VLAN sockets, returning the first
         // available packet with a bonus `vid` tag attached in the metadata.
-        for i in 0..N {
-            let socket = self
-                .get_socket_mut(socket_index, i)
+        for vlan in &mut self.vlan_state {
+            let socket = vlan
+                .get_socket_mut(socket_index)
                 .ok_or(RequestError::Fail(ClientError::BadMessageContents))?;
```

Change made in new tip commit.